### ISSUE TITLE
Plugin marker artifact for precompiled script plugin not configured automatically

### ISSUE THREAD

**Issue by asardaes** [gradle/gradle#15190]
Hello, I posted this question [on Stackoverflow](https://stackoverflow.com/q/64833443/5793905), but if it's indeed a bug, I guess it would be good to have an issue to track it.

Repeating the content from above:

The [example in the documentation](https://docs.gradle.org/current/userguide/custom_plugins.html#sec:precompiled_plugins) assumes the plugin lives in the same project that uses it in the `buildSrc` folder,
however, I want to publish the plugin to an internal repository so that it can be used by other projects.
The [documentation](https://docs.gradle.org/current/userguide/java_gradle_plugin.html#java_gradle_plugin) for Java-based Gradle plugins states that applying the `java-gradle-plugin` will

> Configure the Maven or Ivy Publish Plugins publishing plugins to publish a Plugin Marker Artifact for each plugin.

This doesn't seem to be the case for `groovy-gradle-plugin`. I managed to [manually create the marker](https://gist.github.com/asardaes/6abf1fede590257e3448e59fd0d0af2a), but maybe I'm missing something? I'm using Gradle 6.6 and 6.7 across different repositories.

**Comment by emmanuelguerin** [comment]
Just tried in Gradle 7.5 and reproduced the problem.

To be able to publish the plugin anyway, I had to copy/paste some code in my script:

```
plugins {
    id 'groovy-gradle-plugin'
    id 'maven-publish'
}

private void createMavenMarkerPublication(PluginDeclaration declaration, final MavenPublication coordinates, PublicationContainer publications) {
    String pluginId = declaration.getId();
    String pluginGroupId = coordinates.getGroupId();
    String pluginArtifactId = coordinates.getArtifactId();
    String pluginVersion = coordinates.getVersion();
    def publication = publications.create(declaration.getName() + "PluginMarkerMaven", MavenPublication.class);
    publication.setAlias(true);
    publication.setArtifactId(pluginId + ".gradle.plugin");
    publication.setGroupId(pluginId);
    publication.getPom().withXml(new Action<XmlProvider>() {
        @Override
        public void execute(XmlProvider xmlProvider) {
            def root = xmlProvider.asElement();
            def document = root.getOwnerDocument();
            def dependencies = root.appendChild(document.createElement("dependencies"));
            def dependency = dependencies.appendChild(document.createElement("dependency"));
            def groupId = dependency.appendChild(document.createElement("groupId"));
            groupId.setTextContent(pluginGroupId);
            def artifactId = dependency.appendChild(document.createElement("artifactId"));
            artifactId.setTextContent(pluginArtifactId);
            def version = dependency.appendChild(document.createElement("version"));
            version.setTextContent(pluginVersion);
        }
    });
    publication.getPom().getName().set(declaration.getDisplayName());
    publication.getPom().getDescription().set(declaration.getDescription());
}

afterEvaluate {
    gradlePlugin.plugins.each { pluginDeclaration ->
        createMavenMarkerPublication(pluginDeclaration, publishing.publications.pluginMaven, publishing.publications)
    }
}
```

The problem is how the afterEvalute actions are sequenced between the MavenPluginPublishPlugin and the groovy-gradle-plugin.

**Comment by Jarvvski** [comment]
is there any traction on this?

**Comment by Vampire** [comment]
Yes, this is one of the famous ordering problems you earn when using `afterEvaluate`.
I tried this with Gradle 8.4 and it indeed fails to create the marker publications if you first apply `maven-publish` and after that `groovy-gradle-plugin`.
But if you first apply the `groovy-gradle-plugin` and after that `maven-publish` it works properly.
If you do so and it still does not work, you most probably apply the `maven-publish` plugin already earlier for example as part of some other plugin you apply.

An alternative possibility would be to use Kotlin DSL precompiled script plugins, because the Kotlin DSL variant of the plugin does it correctly and does not have an ordering constraint. Along the way you then also have them type-safe, with actually helpful error messages if you mess up syntax, and with amazingly better IDE support when using a proper IDE like IntelliJ.

**Comment by asardaes** [comment]
For completeness, the documentation has included a [note regarding this](https://docs.gradle.org/current/userguide/custom_plugins.html#note_for_plugins_published_without_java_gradle_plugin) for a while. I personally find the workaround cumbersome if you have many consumers because each one would need to apply the config change, but the manual creation approach I linked to in the original post still works for me...

**Comment by Vampire** [comment]
That note has nothing to do with this issue @asardaes.
The `groovy-gradle-plugin` automatically applies the mentioned `java-gradle-plugin` from that note, so it is present.
This here is purely about the order of registered `afterEvaluate` actions and thus the order of the plugins as I described. :-)

**Comment by ljacomet** [comment]
Sorry for the late reply!

The issue is in the backlog of the relevant team, but the existence of a workaround means it might take a while before a fix is made.

***

As indicated, the workaround is to change the order of plugin applications. But this indeed should be fixed so ordering does not matter.
